nav2_graceful_controller/src/graceful_controller.cpp

```diff
@@ -360,6 +408,43 @@ bool GracefulController::inCollision(const double & x, const double & y, const d
   return false;
 }
 
+void GracefulController::computeDistanceAlongPath(
+  const std::vector<geometry_msgs::msg::PoseStamped> & poses,
+  std::vector<double> & distances)
+{
+  distances.resize(poses.size());
+  // Do the first pose from robot
+  double d = std::hypot(poses[0].pose.position.x, poses[0].pose.position.y);
+  distances[0] = d;
+  // Compute remaining poses
+  for (size_t i = 1; i < poses.size(); ++i) {
+    d += nav2_util::geometry_utils::euclidean_distance(poses[i - 1].pose, poses[i].pose);
+    distances[i] = d;
+  }
+}
+
+void GracefulController::validateOrientations(
+  std::vector<geometry_msgs::msg::PoseStamped> & path)
+{
+  // This really shouldn't happen
+  if (path.empty()) {return;}
+
+  // Check if we actually need to add orientations
+  for (size_t i = 0; i < path.size() - 1; ++i) {
```

Comment:
```suggestion
  for (size_t i = 1; i < path.size() - 1; ++i) {
```

I would see paths having the robot's start pose also populated with the goal but others are not. We don't do that in any nav2 planner to-date, but I could easily imagine it